docs/development/pigs/pig-008.rst

```diff
@@ -0,0 +1,449 @@
+should be used. The obvious entry point is to evaluate one dataset per process and
+join the likelihoods at the end. The current handling of model references
+(the same model instance is passed to different datasets to achieve a generic
+parameter linking by updating the model parameters in place), set's a limitation
+on the parallel evaluation, because Python objects can't easily be shared between
+multiple processes. We are aware of this issue, but propose to solve it later,
+because it does not affect the proposed API for datasets.
+
+
+Dataset serialization
+---------------------
+
+For convenience all dataset classes should support serialization, implemented
+via `.read()` and `.write()` methods. As the dataset as to orchestrate the
+serialization of mutiple objects, such as model, maps, flux-points etc. the best
+option is likely to introduce the serialization with a YAML based index file:
+
+.. code ::
+
+    dataset = MapDataset.read("dataset.yaml") # lazy loading?
+    dataset.write("dataset.yaml")
+
+Where the index file points to the various files needed for initialization of the
+dataset. Here is an example:
+
+.. code::
+
+    dataset:
+        type: map-dataset
+        counts: "obs-123/counts.fits"
+        exposure: "obs-123/exposure.fits"
+        edisp: "obs-123/edisp.fits"
+        psf: "obs-123/psf.fits"
+        model: "model.yaml"
+        background-model: "obs-123/background.fits"
+
+Addtionally one could introduce a single FITS file serializiaton for quickly writing /
+reading datasets to disk.
+
+The `Datasets` object could be serialized equivalently as a list of datasets:
+
+.. code::
+
+    - dataset:
+        type: spectrum-dataset
+        maps:
+            counts: "obs-123/counts.fits"
+            exposure: "obs-123/exposure.fits"
+            edisp: "obs-123/edisp.fits"
+        model: "model.yaml"
+        background-model: "obs-123/background.fits"
+        likelihood: "wstat"
+
+    - dataset:
+        type: spectrum-dataset
+        maps:
+            counts: "obs-124/counts.fits"
+            exposure: "obs-124/exposure.fits"
+            edisp: "obs-124/edisp.fits"
+        model: "model.yaml"
+        background-model: "obs-124/background.fits"
+        likelihood: "wstat"
+
+    - dataset:
+        type: flux-point-dataset
+        flux-points : "fermipy-flux-points.fits"
+        model: "spectral-model.yaml"
+        likelihood: "template"
+
+
+Lazy loading of Datasets
+------------------------
+For the use-case of inspecting individual datasets (e.g. MapDataset per observation)
+it could be advantegeous to implement a lazy-loading or generator interface for
+`Datasets.read()`. Such that the individual datasets are only read from disk on
+requests and are not loaded in memory when calling `.read()`. We propose to handle
+this with a generic lazy-loading mechanism for `Map` objects, where the data is loaded
+on access of the `.data` attribute.
```

Comment:
Really introduce lazy-loading in `Map`?
Not sure it's a good idea, it introduces a lot of complexity, the FITS I/O has to change.
At the DL3 obs level we introduce the lazy-loading at a higher level in `Observations`, not in the EventList, IRF classes.

Suggest to put this as an open question, saying lazy-loading is maybe something we want, e.g. to support parallel evaluation (see paragraph above), but how it will be done isn't decided yet.

Do you discuss the question whether Dataset is used in data reduction as a container?
Doing data reduction for 100s of runs would be another use case where data write and discard from memory is probably a good idea, at least to support optionally.